{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34c2f6ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from sklearn.manifold import TSNE\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import os\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28d1a2ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pairwised data\n",
    "df_result = pd.read_csv('Pairwised_sequences.csv', keep_default_na=False)\n",
    "df_result = df_result[df_result['pairwised_sequence'].str.startswith('M')] # extract the sequences starts from Methionine\n",
    "df_result = df_result[df_result['pairwised_sequence'].str.len() == 1273] # extract the sequences whose length is 1273"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b447e1a",
   "metadata": {},
   "outputs": [],
   "source": [
    " ## prepare for creating the dataset for each approach ##\n",
    "\n",
    "# put each sequence to the list\n",
    "pairwised_seq = df_result[\"pairwised_sequence\"]\n",
    "\n",
    "## amino acid parameters ##\n",
    "\n",
    "    # Amino acid property [amino acid volume, amino acid hydrophilicity].\n",
    "    # Both of these are relative values.\n",
    "\n",
    "aa_properties_dict = {\n",
    "    'A':[-2.90, -1.03], 'R':[2.41, 1.31], 'N':[-0.68, 0.79],\n",
    "    'D':[-0.92, 1.23], 'C':[-1.89, 0.15], 'Q':[0.36, 1.09],\n",
    "    'E':[0.16, 1.28], 'G':[-4.04, 0.01], 'H':[0.83, 1.15],\n",
    "    'I':[0.51, -1.32], 'L':[0.52, -1.40], 'K':[0.92, 1.23],\n",
    "    'M':[0.92, -1.42], 'F':[2.22, -1.47], 'P':[-1.25, -0.64],\n",
    "    'S':[-2.36, 0.38], 'T':[-1.19, 0.28], 'W':[4.28, -0.18],\n",
    "    'Y':[2.75, -0.18], 'V':[-0.65, -1.27], '-':[0, 0]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1677c1b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## dataset for amino acid properties ##\n",
    "\n",
    "amino_acid_properties = [[aa_properties_dict[i] for i in seq] for seq in pairwised_seq]\n",
    "\n",
    "pairwised_volume = [[item[0] for item in sublist] for sublist in amino_acid_properties]\n",
    "\n",
    "pairwised_hydrophilicity = [[item[1] for item in sublist] for sublist in amino_acid_properties]\n",
    "\n",
    "amino_acid_volume_hydrophilicity = [sum([volume, hydro], []) for volume, hydro in zip(pairwised_volume, pairwised_hydrophilicity)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15fb6cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## dataset for one-hot encoding ##\n",
    "\n",
    "one_hot_vec_dict={}\n",
    "\n",
    "for i,key in enumerate(aa_properties_dict):\n",
    "    one_hot_vec=np.zeros(len(aa_properties_dict)) # all values are set to zero\n",
    "    one_hot_vec[i]=1 # change the corresponding values to 1\n",
    "    one_hot_vec_dict[key]=one_hot_vec # this obatained  vector is inserted into empty dictionary\n",
    "\n",
    "# create data with this dictionary\n",
    "\n",
    "one_hot_datas = []\n",
    "for seq in pairwised_seq:\n",
    "    seq_data=[]\n",
    "    \n",
    "    for lit in seq:\n",
    "        seq_data.append(one_hot_vec_dict[lit]) # assign the parameters of one-hot encoding\n",
    "    \n",
    "    one_hot_array = np.array(seq_data) # change the style of array in list to array only\n",
    "    one_hot_datas.append(one_hot_array.reshape(21*len(one_hot_array)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "950b131d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## dataset for k-mer based approach ##\n",
    "\n",
    "# k=3, create the all pattern of three amino acid sequence\n",
    "\n",
    "pattern_size = 21*21*21\n",
    "seq_size = len(pairwised_seq[0])\n",
    "kmer_size = seq_size - 3 + 1\n",
    "\n",
    "pattern_aminoseq = []\n",
    "for key in aa_properties_dict:\n",
    "    pattern = []\n",
    "    pattern.append(key)\n",
    "    for keys in aa_properties_dict:\n",
    "        pattern.append(keys)\n",
    "        for ke in aa_properties_dict:\n",
    "            pattern.append(ke)\n",
    "            pattern_aminoseq.append(key+keys+ke)\n",
    "\n",
    "# dictionalize, assign the ID to eazh sequence\n",
    "\n",
    "amino_pattern_dict={}\n",
    "\n",
    "for i in range(pattern_size):\n",
    "    amino_pattern_dict[str(pattern_aminoseq[i])]=i\n",
    "\n",
    "# k-mer_data\n",
    "data_pattern=[]\n",
    "for seq in pairwised_seq:\n",
    "    zeroseq=np.zeros(pattern_size) # create the array of 0, whose length is equal to the sequence length\n",
    "    \n",
    "    for j in range(len(seq)-3):\n",
    "        zeroseq[amino_pattern_dict[str(seq[j:3+j])]]+=1 # when one sequence appear, add 1 to the corresponded 'zeroseq' at the index \n",
    "        \n",
    "    data_pattern.append(zeroseq/kmer_size) # calculate the appearance rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c728dd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## t-sne ##\n",
    "# along with the tSNE, calculate the runtime\n",
    "\n",
    "tsne = TSNE(n_components = 2, random_state = 0, perplexity = 30, n_iter = 1000)\n",
    "\n",
    "# all apps and softwares were quited except for mi、anaconda and juyter notebook in google chrome以外は終了して計測した。\n",
    "\n",
    "# aa\n",
    "\n",
    "# start measurement of runtime\n",
    "time_sta_aa = time.perf_counter()\n",
    "# t-sne start\n",
    "amino_acid_volume_hydro_tsne = tsne.fit_transform(amino_acid_volume_hydrophilicity)\n",
    "# finish measurement of runtime\n",
    "time_end_aa = time.perf_counter()\n",
    "# calculate the time (s)\n",
    "tim_aa = time_end_aa - time_sta_aa\n",
    "\n",
    "df_result['tsne_x_axis_volume_hydro'] = amino_acid_volume_hydro_tsne[:, 0]\n",
    "df_result['tsne_y_axis_volume_hydro'] = amino_acid_volume_hydro_tsne[:, 1]\n",
    "\n",
    "# one hot\n",
    "\n",
    "# start measurement of runtime\n",
    "time_sta_one_hot = time.perf_counter()\n",
    "# t-sne start\n",
    "seq_datas_tsne = tsne.fit_transform(one_hot_datas)\n",
    "# finish measurement of runtime\n",
    "time_end_one_hot = time.perf_counter()\n",
    "# calculate the time (s)\n",
    "tim_one_hot = time_end_one_hot - time_sta_one_hot\n",
    "\n",
    "df_result['tsne_x_axis_seq_datas'] = seq_datas_tsne[:, 0]\n",
    "df_result['tsne_y_axis_seq_datas'] = seq_datas_tsne[:, 1]\n",
    "\n",
    "# k-mer\n",
    "\n",
    "# start measurement of runtime\n",
    "time_sta_kmer = time.perf_counter()\n",
    "# t-sne start\n",
    "data_pattern_tsne = tsne.fit_transform(data_pattern)\n",
    "# inish measurement of runtime\n",
    "time_end_kmer = time.perf_counter()\n",
    "# calculate the time (s)\n",
    "tim_kmer = time_end_kmer - time_sta_kmer\n",
    "\n",
    "df_result['tsne_x_axis_data_pattern'] = data_pattern_tsne[:, 0]\n",
    "df_result['tsne_y_axis_data_pattern'] = data_pattern_tsne[:, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f588f671",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fig.2, plot the clusterig by t-SNE ##\n",
    "\n",
    "# create the function which sort by the order of Greek letter\n",
    "def sort_by_greek_alphabet(x):\n",
    "    greek_order = ['Alpha', 'Beta', 'Gamma', 'Delta', 'Epsilon', 'Eta', 'Iota', 'Kappa', 'Lambda', 'N/A', 'Omicron', 'Zeta', 'Mu', 'other']\n",
    "    return sorted(x, key=greek_order.index)\n",
    "\n",
    "\n",
    "# funtion to create the plot for each variant\n",
    "def create_scatter_plot_by_species(df, x_data, y_data):\n",
    "    \n",
    "    plt.figure(figsize = (12, 8))\n",
    "    color_datasets = plt.cm.get_cmap(\"tab10\").colors + plt.cm.get_cmap(\"Dark2\").colors\n",
    "    \n",
    "    for rgb, labels in zip(color_datasets, sort_by_greek_alphabet(df['species_name'].unique())):\n",
    "        data = df[df['species_name'] == labels]\n",
    "        plt.scatter(data[x_data], data[y_data], label = labels, color = rgb, s = .75)\n",
    "        \n",
    "    plt.xticks(fontsize = 20)\n",
    "    plt.yticks(fontsize = 20)\n",
    "    plt.xlabel('t-SNE x axis', fontsize = 20, fontweight = 'bold')\n",
    "    plt.ylabel('t-SNE y axis', fontsize = 20, fontweight = 'bold')\n",
    "    plt.legend(fontsize = 16, markerscale = 7, bbox_to_anchor=(1.01, 1))\n",
    "\n",
    "\n",
    "os.makedirs(os.path.join('Results', 'Fig_2'), exist_ok=True)\n",
    "\n",
    "# amino acid properties\n",
    "create_scatter_plot_by_species(df_result, 'tsne_x_axis_volume_hydro', 'tsne_y_axis_volume_hydro')\n",
    "plt.savefig(os.path.join('Results', 'Fig_2', 'Fig_2_Clustering_with_aa_properties_by_t-SNE.png'), dpi = 300, bbox_inches = 'tight')\n",
    "\n",
    "# one-hot encoding\n",
    "create_scatter_plot_by_species(df_result, 'tsne_x_axis_seq_datas', 'tsne_y_axis_seq_datas')\n",
    "plt.savefig(os.path.join('Results', 'Fig_2', 'Fig_2_Clustering_by_one_hot_t-SNE.png'), dpi = 300, bbox_inches = 'tight')\n",
    "\n",
    "# k-mer\n",
    "create_scatter_plot_by_species(df_result, 'tsne_x_axis_data_pattern', 'tsne_y_axis_data_pattern')\n",
    "plt.savefig(os.path.join('Results', 'Fig_2', 'Fig_2_Clustering_k-mer_by_t-SNE.png'), dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c0333e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result.to_csv('after_t-sne.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
